advanced/src/app/core/authentication/authentication.service.ts

```diff
@@ -11,59 +10,154 @@ import { environment } from '@env/environment';
+    this.setOrClearLocalStorage('login_path', uri);
+  }
+
+  private setOrClearLocalStorage(key: string, value: any) {
+    if (value) {
+      localStorage.setItem(key, value as string);
+    } else {
+      localStorage.removeItem(key);
+    }
+  }
+
   /** Authorizes with Auth0 */
   public login(): void {
+    this._loginPath = window.location.pathname;
     this.auth0.authorize();
   }
 
-  /** Logs out by clearing the access and ID tokens */
+  /** Clears application credentials and logs the user out from Auth0 */
   public logout(): void {
-    localStorage.removeItem('access_token');
-    localStorage.removeItem('id_token');
-    localStorage.removeItem('user_id');
-    localStorage.removeItem('expires_at');
-    // navigate to the home route
-    location.reload();
-    this.router.navigate(['/']);
+    this.softLogout();
+    this.auth0.logout({
+      // Tell Auth0 to redirect back to the site root
+      // (the site URL should be added to your Auth0 application's list of Allowed Logout URLs)
+      returnTo: window.location.origin
+    });
+  }
+
+  /** Clears application credentials without logging out from Auth0 session */
+  public softLogout(): void {
+    this._userId = null;
+    this._accessToken = null;
+    this._idToken = null;
+    this._expiresAt = null;
   }
 
   /** Checks whether the current time is past the access token's expiry time */
   public isAuthenticated(): boolean {
-    const expiresAt = JSON.parse(localStorage.getItem('expires_at') || '{}');
-    return new Date().getTime() < expiresAt;
+    if (new Date().getTime() < this._expiresAt) {
+      return true;
+    } else {
+      this.softLogout();
+      return false;
+    }
   }
 
+  /** Attempts to renew the user's access and id tokens */
+  public renewTokens(): void {
+    this.auth0.checkSession({}, (err, authResult) => {
+      if (authResult && authResult.accessToken && authResult.idToken) {
+        this.localLogin(authResult);
+      } else if (err) {
+        console.error(
+          'Failed to renew authentication token. Forcing logout.',
+          err
+        );
+        this.logout();
+      }
+    });
+  }
+
+  /** Parses the current window location's hash fragment as Auth0 credentials, stores the results in local session storage */
   public handleAuthentication(): void {
     this.auth0.parseHash((err: any, authResult: any) => {
-      console.log(authResult);
       if (authResult && authResult.accessToken && authResult.idToken) {
+        this.localLogin(authResult);
         window.location.hash = '';
-        this.setSession(authResult);
-        this.router.navigate(['/advanced']);
+
+        // At this point, we could query environment.graphqlEndpoint for user profile details,
+        // perform any initial caching and/or trigger user notifications as appropriate.
+
+        // Redirect the user back to where they were before logging in
+        this.router.navigate([this._loginPath]);
+        setTimeout(() => {
```

Comment:
Wasn't sure about the need for this `setTimeout` -> `window.location.reload()` - is that specific to Cynthesize?

Reply:
This issue exists because of how modules are initialized in Angular. When using lazy routing, the modules at that route are initialized when the route is activated. Here, the GraphQL module that is used to initialize the Apollo client which will be making requests to Hasura on behalf of our application is initialized as soon as the application comes back from the callback. Now, this Module also uses the JWT key and get's it from the localStorage to init the app. 

Now the problem faced is as soon as the app is initialized from the callback, the module is initialized and the JWT key field is left empty because by that time the JWT key was not set in local storage. 

While working with this issue, I found two solutions for this:

1. We could simply take advantage of Angular's lazy loading feature and use the GraphQL module in another route which would be activated only when the JWT key is set. This doesn't require reloading all of the application but this also doesn't allow us to use the Apollo client when the JWT is not set.

2. Another solution is to reload the app after some time from the callback such that all the modules are initialized again and the set value of JWT is used in the module. Now, this solution is a poor solution I could find but that's only the works because, in an ideal application, we'd need the GraphQL module even when the user is not logged in to communicate to Hasura. 

These are my views and workaround for this, please let me know if you find a better solution for this.

Reply:
I also made a change to the GraphQL module to use ApolloLink middleware when it constructs its http links, so JWTs now only get injected on demand - I believe this has fixed the issue.  Certainly after taking the reload out, looks like we're all good now..  I can log out & in as different users and see their different "to do" lists without the extra reload being necessary.